src/pybind/mgr/dashboard/CMakeLists.txt

```diff
@@ -9,34 +9,71 @@ add_dependencies(tests mgr-dashboard-test-venv)
 
 if(WITH_MGR_DASHBOARD_FRONTEND AND NOT CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64|AARCH64|arm|ARM")
 
-set(mgr-dashboard-nodeenv ${CMAKE_CURRENT_BINARY_DIR}/node-env)
+if(WITH_SYSTEM_NPM)
+  # On FreeBSD we need to preinstall all the tools that are required for building
+  # dashboard, because versions fetched are not working on FreeBSD.
+  add_custom_command(
+    OUTPUT "${CMAKE_SOURCE_DIR}/src/pybind/mgr/dashboard/frontend/node_modules"
+    COMMAND ${NPM} install 
+    DEPENDS frontend/package.json
+    WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}/src/pybind/mgr/dashboard/frontend
+    COMMENT "dashboard frontend dependencies are being installed"
+  )
```

Comment:
please avoid repeating these `add_custom_command()` calls in both if branches.

Reply:
@tchaikov 
I don't really see blocks that are equal.
Only way to do that is create extra variables in the if-then-else and extract the `custom_command()`
Is rthat what you mean?

Reply:
for example, https://github.com/ceph/ceph/pull/22562/files#diff-e39dffd2b7f8e7e205f4075fff60743eR72 and https://github.com/ceph/ceph/pull/22562/files#diff-e39dffd2b7f8e7e205f4075fff60743eR76 repeat their counterparts at https://github.com/ceph/ceph/pull/22562/files#diff-e39dffd2b7f8e7e205f4075fff60743eR28 . 

anyway, i just posted #22636 in hope to help with this issue. could you help review it?